### PR TITLE
Refactor `PersonalDetails_Update` with a few new commands

### PR DESCRIPTION
**Pull request by Beamanator**
### Fixed Issues

$ https://github.com/Expensify/Expensify/issues/211813

### Tests
Same tests as: https://github.com/Expensify/Web-Expensify/pull/34219:

### `UpdateProfile`
1. Go to profile, update name & timezone, save
2. Make sure everything updates correctly (& in multiple devices!)
3. Verify you see this in the App logs:
```
[info] [Report] Handled onyxApiUpdate event sent by Pusher - [{"onyxMethod":"merge","key":"personalDetails","value":{"alexbeaman@expensifail.com":{"firstName":"TheE","lastName":"OriginalFake","pronouns":"__predefined_callMeByMyName"}}}]
[info] [Report] Handled onyxApiUpdate event sent by Pusher - [{"onyxMethod":"merge","key":"personalDetails","value":{"alexbeaman@expensifail.com":{"timezone":{"automatic":false,"selected":"Africa/Cairo"}}}}]
```

### `OpenProfile`
1. Make sure your timezone is set and is ready to update automatically
2. In Onyx, manually change your timezone to a valid timezone, in a different time zone
3. Open profile page
6. Verify Tz updates back to your current timezone everywhere!
5. Verify you see this in the App logs: `[info] [Report] Handled onyxApiUpdate event sent by Pusher - [{"onyxMethod":"merge","key":"personalDetails","value":{"alexbeaman@expensifail.com":{"timezone":{"automatic":true,"selected":"Africa/Cairo"}}}}]`

### `DeleteUserAvatar`
1. Delete your personal user avatar
2. Verify you now get a default avatar set on all devices!
4. Verify you see this in the App logs: `[info] [Report] Handled onyxApiUpdate event sent by Pusher - [{"onyxMethod":"merge","key":"personalDetails","value":{"alexbeaman@expensifail.com":{"avatar":"https://d2k5nsl2zxldvw.cloudfront.net/images/avatars/avatar_7.png"}}}]`

- [x] Verify that no errors appear in the JS console

#### Contributor (PR Author) Checklist
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [ ] I added steps for local testing in the `Tests` section
    - [ ] I added steps for Staging and/or Production testing in the `QA steps` section
    - [ ] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/docs/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I ran the tests on **all platforms** & verified they passed on:
    - [ ] iOS / native
    - [ ] Android / native
    - [ ] iOS / Safari
    - [ ] Android / Chrome
    - [ ] MacOS / Chrome
    - [ ] MacOS / Desktop
- [x] I verified there are no console errors (if there’s a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/docs/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained “why” the code was doing something instead of only explaining “what” the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named “index.js”. All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/docs/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/docs/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] Any functional components have the `displayName` property
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose and it is
- [x] If a new CSS style is added I verified that:
    - [ ] A similar style doesn’t already exist
    - [ ] The style can’t be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.

<details>
<summary><h4>PR Reviewer Checklist</h4></summary>

- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/docs/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] iOS / native
    - [ ] Android / native
    - [ ] iOS / Safari
    - [ ] Android / Chrome
    - [ ] MacOS / Chrome
    - [ ] MacOS / Desktop
- [ ] I verified there are no console errors (if there’s a console error not related to the PR, report it or open an issue for it to be fixed)
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/docs/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained “why” the code was doing something instead of only explaining “what” the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named “index.js”. All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/docs/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/docs/PR_REVIEW_GUIDELINES.md)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] Any functional components have the `displayName` property
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose and it is broken down into smaller components in order to separate concerns and functions
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn’t already exist
    - [ ] The style can’t be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.

</details>

### QA Steps
Same as "Tests" section, but don't worry about checking the logs

- [ ] Verify that no errors appear in the JS console

#### Web & Desktop

Updating timezone by opening profile:

https://user-images.githubusercontent.com/3885503/179777857-6e73d00e-781d-42c3-8ef2-7bd8efd4dafc.mov

Deleting profile image:

https://user-images.githubusercontent.com/3885503/179778307-997c1537-8ce1-4154-896e-789fd1b320a8.mov

Updating profile info

https://user-images.githubusercontent.com/3885503/179778819-09229aba-067b-4191-ace5-787f36e49b05.mov

### CODE REVIEW

File: src/libs/actions/PersonalDetails.js
Comment (Beamanator) [inline src/libs/actions/PersonalDetails.js:407]: Probably also need to update `avatarHighResolution` - see https://github.com/Expensify/App/issues/9418#issuecomment-1168518507

**Comment by Beamanator** [comment]
The App PR we were waiting on has been merged into `main` & into this branch! Ready for review 👍

**Comment by Gonals** [comment]
Seems like Travis is not happy 😁

**Comment by Gonals** [comment]
This should be on HOLD until the Web counterpart merges, right?

**Comment by Beamanator** [comment]
3 videos uploaded!

File: src/pages/settings/Profile/ProfilePage.js
Comment (Beamanator) [inline src/pages/settings/Profile/ProfilePage.js:103]: Hmm one thought here: I could just update the state in `removeAvatar` instead of checking for avatar changes here 🤔

Comment (Beamanator) [reply src/pages/settings/Profile/ProfilePage.js]: I decided to do my idea - remove this from `componentDidMount` because I don't want to have to check avatar changes on EVERY re-render, instead we can immediately update state to default avatar when deleting the user's avatar

**Comment by Beamanator** [comment]
Ok fixed conflicts & made 1 tiny change

Web-E PR has been merged (not deployed to prod yet)

**Comment by Beamanator** [comment]
One more thing we can test with this PR which I think we should make sure is working: https://expensify.slack.com/archives/C01GTK53T8Q/p1658938208861039

File: src/libs/actions/PersonalDetails.js
Comment (Gonals) [inline src/libs/actions/PersonalDetails.js:268]: This would require changes on the Web side too, but it would be better to have these parameters separated. That way, we'll have better `_inputRules` checking on each of them (I just got this comment on my PR 😆)

File: src/pages/settings/Profile/ProfilePage.js
Comment (Gonals) [inline src/pages/settings/Profile/ProfilePage.js:70]: We can use `lodashGet` for these

**Review by Gonals — COMMENTED** [review]
A couple of comments!

File: src/libs/actions/App.js
Comment (mountiny) [inline src/libs/actions/App.js:221]: :sus: if `oldTimezoneData.automatic` is `false`, this will be `true`, it it was `true`, this will be `true`, right? Is this how we want it? we should probably refactor

File: src/pages/settings/Profile/ProfilePage.js
Comment (Beamanator) [reply src/pages/settings/Profile/ProfilePage.js:70]: True I will update!

Comment (mountiny) [reply src/pages/settings/Profile/ProfilePage.js:70]: +1

File: src/pages/settings/Profile/ProfilePage.js
Comment (mountiny) [inline src/pages/settings/Profile/ProfilePage.js:160]: NAB: Should we maybe rename this to `deleteAvatar` too to keep it consistent?

**Review by mountiny — COMMENTED** [review]
Nice one, left couple of comments, I havent tested yet!

File: src/libs/actions/PersonalDetails.js
Comment (Beamanator) [reply src/libs/actions/PersonalDetails.js:268]: Good idea! Made an issue for this https://github.com/Expensify/Expensify/issues/220321

Hopefully we can get this merged quicker (once I address other comments) then I'll come back to this refactor 👍

File: src/pages/settings/Profile/ProfilePage.js
Comment (Beamanator) [reply src/pages/settings/Profile/ProfilePage.js:70]: What do y'all think about extracting `this.props.currentUserPersonalDetails` in the `constructor` to something like this:

```js
        const currentUserDetails = this.props.currentUserPersonalDetails || {};
        this.state = {
            firstName: currentUserDetails.firstName || '',
            hasFirstNameError: false,
            lastName: currentUserDetails.lastName || '',
            hasLastNameError: false,
            pronouns: currentUserDetails.pronouns || '',
            hasPronounError: false,
            hasSelfSelectedPronouns: !_.isEmpty(currentUserDetails.pronouns) && !currentUserDetails.pronouns.startsWith(CONST.PRONOUNS.PREFIX),
            selectedTimezone: lodashGet(currentUserDetails, 'timezone.selected', CONST.DEFAULT_TIME_ZONE.selected),
            isAutomaticTimezone: lodashGet(currentUserDetails, 'timezone.automatic', CONST.DEFAULT_TIME_ZONE.automatic),
            logins: this.getLogins(props.loginList),
            avatar: {uri: currentUserDetails.avatar || ReportUtils.getDefaultAvatar(currentUserDetails.login)},
            isAvatarChanged: false,
        };
```

Comment (Beamanator) [reply src/pages/settings/Profile/ProfilePage.js:70]: I like ^ better b/c:
1. We only have to use `lodashGet` where we're accessing nested user detail properties (`timezone.selected` & `timezone.automatic`)
2. It makes some lines a lot shorter, where we were doing `lodashGet(this.props.currentUserPersonalDetails, 'something')` multiple times

File: src/pages/settings/Profile/ProfilePage.js
Comment (Beamanator) [reply src/pages/settings/Profile/ProfilePage.js:160]: I think I'm down with this :D

File: src/pages/settings/Profile/ProfilePage.js
Comment (Beamanator) [reply src/pages/settings/Profile/ProfilePage.js:70]: This can also help us here, where we'd normally want to use `lodashGet` everywhere: https://github.com/Expensify/App/blob/f6736b1a1fffdab3893bfb152dfe0be331d57bb7/src/pages/settings/Profile/ProfilePage.js#L212-L217

File: src/libs/actions/App.js
Comment (Beamanator) [reply src/libs/actions/App.js:221]: Aah trueee I was thinking `??` but I don't think we like using that in our app code 😅 good catch :D

I think I can use `lodashGet` here like this:

Screen Shot 2022-07-28 at 11 50 14 AM

File: src/pages/settings/Profile/ProfilePage.js
Comment (mountiny) [reply src/pages/settings/Profile/ProfilePage.js:70]: @Beamanator I am fine with that proposal, nice!

File: src/libs/actions/App.js
Comment (mountiny) [reply src/libs/actions/App.js:221]: What is the automatic actually? Isn't this automatic all the time when we call open profile and get the timezone from `moment.guess`? Should it not be marked as `automatic: true` in this case everytime?

So many questions haha

Comment (Beamanator) [reply src/libs/actions/App.js]: Soooooooo `automatic` is the setting if the user wants to have the timezone update automatically or not.

The related Web-E logic is here: https://github.com/Expensify/Web-Expensify/blob/ae9d6761900b4518c1f10106ef61e27138362b97/lib/UserAPI.php#L3227-L3233

Basically, we get the current timezone NVP value. If the timezone's `automatic` setting is `true`, we update the NVP to whatever the new timezone value is. If it's `false`, we don't do anything

File: src/pages/settings/Profile/ProfilePage.js
Comment (Beamanator) [reply src/pages/settings/Profile/ProfilePage.js]: Committed!

File: src/libs/actions/App.js
Comment (mountiny) [inline src/libs/actions/App.js:223]: @Beamanator Ok, so you will pass in the new one even if we do not want it automatically updated and in that case this just wont be saved in Web-E?

Comment (Beamanator) [reply src/libs/actions/App.js:223]: yeah exactly

**Comment by Beamanator** [comment]
Responded to comments, fixed a few things, refactored a few things, ready for review again :D

**Review by Gonals — DISMISSED** [review]
LGTM!

**Comment by mountiny** [comment]
Trying to test

**Comment by mountiny** [comment]
image
getting this testing the first scenario wiht a fresh account but this seems like if we set the timezone weird on first sign up

File: src/libs/actions/PersonalDetails.js
Comment (mountiny) [inline src/libs/actions/PersonalDetails.js:282]: @Beamanator should this optimistic data actually update it in the UI automatically? Is it that we do not rerender the RHS when we go back from the settings? 

In profile settings:
image

Clicking back to the settings, still old name 
image


I think that should be fixed because as user I would be perplexed what is correct.

**Review by mountiny — CHANGES_REQUESTED** [review]

**Comment by mountiny** [comment]
No, thank you for working on this!

**Comment by Beamanator** [comment]
Ok @vitHoracek it looks like this is the fix you're looking for, right? I'm about to commit these changes

https://user-images.githubusercontent.com/3885503/181746621-2ab84600-db49-4459-a1f6-33b99dffed38.mov

**Comment by mountiny** [comment]
@Beamanator YAAASS

File: src/libs/actions/PersonalDetails.js
Comment (Beamanator) [inline src/libs/actions/PersonalDetails.js:283]: FYI this will need to be added in Web-E, I plan to add this requirement in this issue: https://github.com/Expensify/Expensify/issues/220321

**Comment by Beamanator** [comment]
Great!! This will not update on all clients yet (see my previous comment) but I'll fix that in a following Web-E PR 👍

File: src/libs/actions/PersonalDetails.js
Comment (mountiny) [reply src/libs/actions/PersonalDetails.js:283]: Ok, but we should not get any problems when sending this. It will get updated for other users when their refresh, but it is not included in the Pusher update you saying?

Comment (Beamanator) [reply src/libs/actions/PersonalDetails.js:283]: Right right exactly @mountiny - nothing will break, only the client making the change will get these updates automatically, others will get it after refresh 👍

**Comment by Beamanator** [comment]
Actually before we merge this, I'm going to add `firstName`, `lastName`, and `pronouns` as their own params to `UpdateProfile` so that it will be fewer PRs to change from `details` to the individual params

File: src/libs/actions/PersonalDetails.js
Comment (Beamanator) [reply src/libs/actions/PersonalDetails.js:269]: Adding `firstName`, `lastName`, `pronouns` below so my next Web-E PR can accept those & not care about `details` anymore

**Comment by Beamanator** [comment]
Ok now I think we're good to go again

File: src/libs/actions/App.js
Comment (mountiny) [inline src/libs/actions/App.js:222]: Discussed with Alex in DM, I think this function does not work as expected with the VPNs and it does not set you to the timezone of the area. However, that is how we have it now and I would argue it is better as in the chat you want others to see your actual timezone and not the VPN timezone.

File: src/pages/settings/Profile/ProfilePage.js
Comment (mountiny) [inline src/pages/settings/Profile/ProfilePage.js:77]: @Beamanator I think this change is not gonna result in the same boolean right? 

if pronouns is empty, we will get false for `hasSelfSelectedPronouns ` here:

`!_.isEmpty(this.props.currentUserPersonalDetails.pronouns)`

but if pronouns is empty with the new line, that would be startsWith(PREFIX) is false and negating the entire line will give `true`.

Nonetheless, even if it is correct, it is hard to understand 😄

**Review by mountiny — COMMENTED** [review]
I have one more outstanding question there.

Otherwise I think we are good to go with how the timezones work now, but we should update the test steps. Also I would copy them over from the Web-E so it is easier for QA to follow!

Great job Alex, this one is now easy

File: src/pages/settings/Profile/ProfilePage.js
Comment (Beamanator) [reply src/pages/settings/Profile/ProfilePage.js:77]: Dude good shout, this would probably cause weird behavior if `pronouns` is empty 🙃 👍 

I can revert part of the refactor here, just keeping the `currentPersonalDetails` simplification

Comment (mountiny) [reply src/pages/settings/Profile/ProfilePage.js:77]: @Beamanator Sounds good!

**Review by mountiny — APPROVED** [review]
Thanks for working on thins and for all the changes!

**Comment by Beamanator** [comment]
Added necessary check back (for has personal pronouns), brought tests here from the Web-E PR, and changed the timezone checks to not test by using a VM

**Comment by mountiny** [comment]
Since Alberto approved the changes before, let's :shipit:

File: src/pages/settings/Profile/ProfilePage.js
Comment (Beamanator) [inline src/pages/settings/Profile/ProfilePage.js:191]: This tiny change gave rise to https://github.com/Expensify/App/issues/10215

Note: The main issue is opacity being incorrect, not the Growl being missing

**Pull request merged**
